modules/formats/FDB.py

```diff
@@ -42,32 +45,99 @@ def open_command(self, f):
 		with open(command_path, "r") as file:
 			return file.read()
 
+	@staticmethod
+	def hash_djb2(s):
+		hash = 5381
+		for x in s:
+			hash = (( hash << 5) + hash) + ord(x)
+		return hash & 0xFFFFFFFF
+
+	ScriptContext = namedtuple('ScriptContext', ['name', 'command', 'num_strings', 'find_strings'])
+	def context_is_valid(self, context: ScriptContext, name_tuples):
+		if not context or not name_tuples:
+			return False
+		# Require at least N tuples from the GUI, strings over context.num_strings are ignored
+		if len(name_tuples) < context.num_strings:
+			logging.warning(f"Required replacement strings missing for {self.file_entry.name}")
+			logging.info(f"Skipping {self.file_entry.name}. Script strings in need of replacement: {context.find_strings}")
+			logging.info(f"Rename Contents on this FDB needs {context.num_strings} strings in both Find/Replace separated by new lines.")
+			return False
+		return context.name and context.command and context.find_strings
+
 	def rename_content(self, name_tuples):
-		command = None
-		s = None
+		# The current script context e.g. "animals" or "research"
+		context = None
+		# The SQL strings per script
+		script_strings = {
+			"animals" : [("ORIGINAL_SPECIES", "NEW_SPECIES"), ("ORIGINAL_PREFAB", "NEW_PREFAB")],
+			"education" : [("ORIGINAL_SPECIES", "NEW_SPECIES")],
+			"research" : [("ORIGINAL_SPECIES", "NEW_SPECIES")],
+			"zoopedia": [("ORIGINAL_SPECIES", "NEW_SPECIES")]
+		}
+
 		if is_pz(self.ovl) or is_pz16(self.ovl):
 			for s in ("zoopedia", "research", "education", "animals"):
 				if s in self.file_entry.name:
-					command = self.open_command(f"pz_{s}")
+					# The SQL strings for the current context
+					find_strings = list(script_strings.get(s, [()]))
+					context = self.ScriptContext(s, self.open_command(f"pz_{s}"), len(find_strings), find_strings)
 					break
-		if command:
+
+		if self.context_is_valid(context, name_tuples):
 			logging.info(f"Executing command '{s}' on {self.file_entry.name}")
 			try:
 				temp_dir, out_dir_func = self.get_tmp_dir()
 				fdb_path = self.extract(out_dir_func)[0]
-				con = sqlite3.connect(fdb_path)
-				cur = con.cursor()
-
-				for old, new in name_tuples:
-					command_replaced = command.replace("ORIGINAL", old).replace("NEW", new)
-					# print(command_replaced)
-					cur.executescript(command_replaced)
-				# Save (commit) the changes
-				con.commit()
-				con.close()
+
+				# Clean up with VACUUM first
+				try:
+					con = sqlite3.connect(fdb_path)
+					con.execute("VACUUM")
+				except sqlite3.Error as e:
+					logging.error(f"SQL error: {str(e)}")
+				finally:
+					con.close()
+
+				# Before running SQL commands, verify the strings exist or you will get empty FDBs
+				run = True
+				with open(fdb_path, 'rb', 0) as file, mmap.mmap(file.fileno(), 0, access=mmap.ACCESS_READ) as s:
```

Comment:
why use mmap here?
maybe pull out into function check_if_str_in_fdb() for better readability